relay-config/src/config.rs

```diff
@@ -11,39 +11,61 @@ use failure::{Backtrace, Context, Fail};
 use serde::{de::DeserializeOwned, Deserialize, Serialize};
 
 use relay_auth::{generate_key_pair, generate_relay_id, PublicKey, RelayId, SecretKey};
-use relay_common::Dsn;
+use relay_common::{Dsn, Uuid};
 use relay_redis::RedisConfig;
 
 use crate::types::ByteSize;
 use crate::upstream::UpstreamDescriptor;
+use std::str::FromStr;
 
 macro_rules! ctry {
     ($expr:expr, $kind:expr, $path:expr) => {
         match $expr {
             Ok(val) => val,
-            Err(err) => return Err(ConfigError::new($path, err.context($kind))),
+            Err(err) => return Err(ConfigError::new_file_error($path, err.context($kind))),
         }
     };
 }
 
+/// Defines the source of a config error
+#[derive(Debug)]
+enum ErrorSource {
+    /// an error originating from a configuration file
+    File(PathBuf),
+    /// an error originating in a field override (an Env. Var. or a CLI parameter)
+    FieldOverride { name: String, value: String },
+}
+
 /// Indicates config related errors.
 #[derive(Debug)]
 pub struct ConfigError {
-    filename: PathBuf,
+    error_source: ErrorSource,
     inner: Context<ConfigErrorKind>,
 }
 
 impl ConfigError {
-    fn new<P: AsRef<Path>>(p: P, inner: Context<ConfigErrorKind>) -> Self {
+    fn new_file_error<P: AsRef<Path>>(p: P, inner: Context<ConfigErrorKind>) -> Self {
         ConfigError {
-            filename: p.as_ref().to_path_buf(),
+            error_source: ErrorSource::File(p.as_ref().to_path_buf()),
             inner,
         }
     }
-
-    /// Returns the filename that failed.
-    pub fn filename(&self) -> &Path {
-        &self.filename
+    fn new_field_error<NameT, ValueT>(
+        name: NameT,
+        value: ValueT,
+        inner: Context<ConfigErrorKind>,
+    ) -> Self
+    where
+        NameT: Into<String>,
+        ValueT: Into<String>,
```

Comment:
It seems that the field name is (and should) always be a `&'static str`. Values are always effectively strings. Can we simplify this signature then?

I'm wondering if we actually need the original value. The inner error will usually contain sufficient information to see what the problem is. Especially for values like the public/private keys, it may not make sense to print them out again. Personally, I would just retain the field name an inner error.

Finally, since this type is an error, so the function name can be simplified.

```rust
fn for_field(name: &'static str, value: String, ...)
```

Reply:
ok